libs/pilight/events/actions/pushsafer.c

```diff
@@ -0,0 +1,225 @@
+	struct JsonNode *jicon = NULL;
+	struct JsonNode *jsound = NULL;
+	struct JsonNode *jvibration = NULL;
+	struct JsonNode *jprivatekey = NULL;
+	struct JsonNode *jvalues = NULL;
+	struct JsonNode *jchild = NULL;
+	int nrvalues = 0;
+
+	jtitle = json_find_member(obj->arguments, "TITLE");
+	jmessage = json_find_member(obj->arguments, "MESSAGE");
+	jprivatekey = json_find_member(obj->arguments, "PRIVATEKEY");
+	jdevice = json_find_member(obj->arguments, "DEVICE");
+	jicon = json_find_member(obj->arguments, "ICON");
+	jsound = json_find_member(obj->arguments, "SOUND");
+	jvibration = json_find_member(obj->arguments, "VIBRATION");
+
+	if(jmessage == NULL) {
+		logprintf(LOG_ERR, "pushsafer action is missing a \"MESSAGE\"");
+		return -1;
+	}
+	if(jprivatekey == NULL) {
+		logprintf(LOG_ERR, "pushsafer action is missing a \"PRIVATEKEY\"");
+		return -1;
+	}
+	nrvalues = 0;
+	if((jvalues = json_find_member(jmessage, "value")) != NULL) {
+		jchild = json_first_child(jvalues);
+		while(jchild) {
+			nrvalues++;
+			jchild = jchild->next;
+		}
+	}
+	if(nrvalues != 1) {
+		logprintf(LOG_ERR, "pushsafer action \"MESSAGE\" only takes one argument");
+		return -1;
+	}
+	nrvalues = 0;
+	if((jvalues = json_find_member(jprivatekey, "value")) != NULL) {
+		jchild = json_first_child(jvalues);
+		while(jchild) {
+			nrvalues++;
+			jchild = jchild->next;
+		}
+	}
+	if(nrvalues != 1) {
+		logprintf(LOG_ERR, "pushsafer action \"PRIVATEKEY\" only takes one argument");
+		return -1;
+	}
+	return 0;
+}
+
+static void *thread(void *param) {
+	struct rules_actions_t *pth = (struct rules_actions_t *)param;
+	struct JsonNode *arguments = pth->parsedargs;
+	struct JsonNode *jtitle = NULL;
+	struct JsonNode *jmessage = NULL;
+	struct JsonNode *jdevice = NULL;
+	struct JsonNode *jicon = NULL;
+	struct JsonNode *jsound = NULL;
+	struct JsonNode *jvibration = NULL;
+	struct JsonNode *jprivatekey = NULL;
+	struct JsonNode *jvalues1 = NULL;
+	struct JsonNode *jvalues2 = NULL;
+	struct JsonNode *jvalues3 = NULL;
+	struct JsonNode *jvalues4 = NULL;
+	struct JsonNode *jvalues5 = NULL;
+	struct JsonNode *jvalues6 = NULL;
+	struct JsonNode *jvalues7 = NULL;
+	struct JsonNode *jval1 = NULL;
+	struct JsonNode *jval2 = NULL;
+	struct JsonNode *jval3 = NULL;
+	struct JsonNode *jval4 = NULL;
+	struct JsonNode *jval5 = NULL;
+	struct JsonNode *jval6 = NULL;
+	struct JsonNode *jval7 = NULL;
+
+	action_pushsafer->nrthreads++;
+
+	char url[1024], typebuf[70];
+	char *data = NULL, *tp = typebuf;
+	int ret = 0, size = 0;
+
+	jtitle = json_find_member(arguments, "TITLE");
+	jmessage = json_find_member(arguments, "MESSAGE");
+	jprivatekey = json_find_member(arguments, "PRIVATEKEY");
+	jdevice = json_find_member(arguments, "DEVICE");
+	jicon = json_find_member(arguments, "ICON");
+	jsound = json_find_member(arguments, "SOUND");
+	jvibration = json_find_member(arguments, "VIBRATION");
+
+	if(jmessage != NULL && jprivatekey != NULL) {
+		jvalues1 = json_find_member(jtitle, "value");
```

Comment:
You should test your own code with all kinds of argument combination and check if it works. Also check for invalid combinations and check if you don't get errors. These tests (and their solutions) should be part of every developers workflow.